monix-tail/shared/src/main/scala/monix/tail/internal/IterantFoldWhileLeft.scala

```diff
@@ -45,26 +47,30 @@ private[tail] object IterantFoldWhileLeft {
       if (hasResult)
         stop.map(_ => s)
       else
-        rest.flatMap(loop(s))
+        rest.flatMap(loop(stopRef)(s))
     }
 
-    def loop(state: S)(self: Iterant[F, A]): F[S] = {
+    def loop(stopRef: ObjectRef[F[Unit]])(state: S)(self: Iterant[F, A]): F[S] = {
       try self match {
         case Next(a, rest, stop) =>
+          stopRef.elem = stop
           f(state, a) match {
-            case Left(s) => rest.flatMap(loop(s))
+            case Left(s) => rest.flatMap(loop(stopRef)(s))
             case Right(s) => stop.map(_ => s)
           }
 
         case NextCursor(cursor, rest, stop) =>
-          process(state, cursor, rest, stop)
+          stopRef.elem = stop
+          process(stopRef)(state, cursor, rest, stop)
 
         case NextBatch(batch, rest, stop) =>
+          stopRef.elem = stop
           val cursor = batch.cursor()
-          process(state, cursor, rest, stop)
+          process(stopRef)(state, cursor, rest, stop)
 
-        case Suspend(rest, _) =>
-          rest.flatMap(loop(state))
+        case Suspend(rest, stop) =>
+          stopRef.elem = stop
+          rest.flatMap(loop(stopRef)(state))
 
         case Last(a) =>
           F.pure(f(state, a) match {
```

Comment:
Calling `f` can throw, which can trigger an exception that's going to trigger our `handleErrorWith`, which is going to call an `earlyStop` reference that's no longer in service.

Do `stopRef.elem = null`.

Reply:
I see, I've fixed it and added tests for this case - please see if I understand desired behavior correctly. :)